### ISSUE TITLE
Finite differences performance

### ISSUE THREAD

**Issue by AxlaDesign** [SciML/NeuralPDE.jl#152]
Going back to this thread:
https://discourse.julialang.org/t/diffeqflux-autodifferentiating-inside-loss-function/47203

I've been cannibalizing some code from NeuralPDE to perform numerical differentiation inside a loss function for scientific machine learning:

```julia
function get_derivative_simplified(order,phi,x,θ)

    _derivative = (x,θ,order) ->
    begin
        if order == 1
            return (phi(x+cbrt(eps(Float32)),θ)[1] - phi(x-cbrt(eps(Float32)),θ)[1])/(2*cbrt(eps(Float32)))

        else
            return (_derivative(x+cbrt(eps(Float32)),θ,order-1)
                  - _derivative(x-cbrt(eps(Float32)),θ,order-1))/(2*cbrt(eps(Float32)))
        end

    end

    return _derivative(x,θ,order)
end

```
And then in the loss function:
```julia
function loss(p)    
    df(x)=get_derivative_simplified(1,NN,x,p)
    du_NN = [df(x) for x in collect(xsteps)]
    mean(broadcast(-, du_true, du_NN))
end
```

This approach gave a good result, but made the whole training take approximately 30% longer than just explicitly writing the finite differences inside the loss function:

```julia
    function loss(p)
    f(x) = NN(x, p)[1]    
    df(x) = (f(x + cbrt(eps(Float32))) - f(x - cbrt(eps(Float32)))) / (2*cbrt(eps(Float32)))
    du_NN = [df(x) for x in collect(xsteps)]
    mean(broadcast(-, du_true, du_NN))
end
```

For my actual project (where I also need second order derivatives), the difference was rather 50%. My guess is that this is due to the recursive and conditional stuff going around in the `get_derivative_simplified` function. 

@ChrisRackauckas comment:

"It’ll depend on the neural network size, with larger NNs seeing less overhead. This overhead can be completely removed though by using an iterator form for the loss function instead of building a bunch of arrays, and making sure everything is concretely typed. I’ll talk with @kirillzubov about it."

**Comment by ChrisRackauckas** [comment]
Yeah, it's these untyped containers that are killing the performance: https://github.com/SciML/NeuralPDE.jl/blob/master/src/pinns_pde_solve.jl#L117 they need to be replaced with `map` calls.
